### DISCUSSION TITLE
Experimental Features

### DISCUSSION THREAD

**Discussion by Kludex** [Kludex/uvicorn#1762] [Ideas] [answered]
Hi there :wave: 

There are many features that we reject because of complexity it can bring to `uvicorn` itself. I personally would like to support them, but I don't want to be too disruptive with the ideas of being "minimal" that `uvicorn`, and `encode` itself tries to build things.

For that reason, I'd like to propose a way to implement those features, at least the ones that make sense, even if they bring complexity. The goal is to make features available, and study if they should be included in `uvicorn` "core" itself.

## What features are we talking about?

To list some:
- Support for ASGI HTTP Trailers extension.
- Support for ASGI TLS extension.
- Support for ASGI Websocket Denial Response extension.
- Support Zero Copy Send extension.
- Support HTTP/2.
- Support a way to customize the access logger.
- Multiprocess class with monitoring.

## What's the alternative here?

I can implement my own HTTP class, and use with `uvicorn`, as the `http` parameter accepts a class, and document that on `uvicorn`.

Doing the alternative approach, I'll lose the potential to understand our users' needs.

## How do you think this experimental approach will work?

This can be thought a bit better, but the approach that I thought was to create a module called `experimental`, and implement the new HTTP class there. If the user wants to use it, they will need to do `uvicorn --http uvicorn.experimental.HTTPProtocol`.

I'd also thought about sending a log message when people use this experimental feature, asking for feedback to understand their use-case, with a permanent discussion on GitHub to receive those feedbacks. 

Thoughts about this? @encode/maintainers

**Comment by iudeen** [comment]
I like the idea of having experimental features rolled to enthusiast users. 

`uvicorn --http uvicorn.experimental.HTTPProtocol` is a neat way for the users to enable such features. Alternatively, we may roll this out as separate package, something like plugins. 

Also, I personally feel a lot of users would benefit from support for ASGI Websocket Denial Response extension and support for HTTP/2.

**Reply by lovelydinosaur** [reply]
> Alternatively, we may roll this out as separate package, something like plugins.

That's a pretty interesting idea.

I like what @florimondmanca did with his https://github.com/florimondmanca/httpx-sse package.
Doesn't necessarily need to be up on PyPI.
A well presented repo and pip installing from GitHub could be a good starting point.

**Comment by abersheeran** [comment]
Yes, the plugin way is good. It can reduce maintenance costs and enrich the functions of uvicorn.

**Comment by Kludex** [comment, marked as answer]
Very well. I'll be organizing this new repository to fit the list of features.

Thank you for the thoughts. :pray:

**Reply by Kludex** [reply]
I've started my work here: https://github.com/Kludex/uvicorn-extensions.
